### PR TITLE
feat(dataframe): implement search related Scalar, DataFrame and Series

### PR DESCRIPTION
**Pull request by AChenQ**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-313b6uy 编写不同类型sql.Series，支持将所有操作转为运算公式](https://app.clickup.com/t/313b6uy)

File: graviti/dataframe/sql/container.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import graviti.portex as pt
+
+if TYPE_CHECKING:
+    from graviti.portex.base import PortexType
+
+
+_T = TypeVar("_T", bound="SearchContainer")
+
+
+class SearchContainer:
```
Comment (linjiX) [inline graviti/dataframe/sql/container.py:19]: seems this class is useless?

File: graviti/dataframe/sql/series.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        self.expr = expr
+        self.schema = schema
+
+    def __eq__(self, right: Any) -> "BooleanSeries":  # type: ignore[override]
+        if isinstance(right, SeriesBase):
+            expr = {"$eq": [self.expr, right.expr]}
+        else:
+            expr = {"$eq": [self.expr, right]}
+
+        return BooleanSeries(expr, pt.boolean())
```
Comment (linjiX) [inline graviti/dataframe/sql/series.py:41]: the type of BooleanSeries must be pt.boolean(), do we need to set type as a parameter?

File: graviti/dataframe/sql/series.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    @classmethod
+    def _get_numeric_operator(cls: Type[_S], opt: str) -> Callable[..., _S]:
+        def func(left: _S, right: Any) -> _S:
+            if isinstance(right, SeriesBase):
+                cls.check_type_for_right(left, right)
+                expr = {f"${opt}": [left.expr, right.expr]}
+            else:
+                expr = {f"${opt}": [left.expr, right]}
+
+            return cls(expr, left.schema)
```
Comment (linjiX) [inline graviti/dataframe/sql/series.py:88]: we except `int + float = float`, but here `int + float = int`

**Review by linjiX — CHANGES_REQUESTED** [review]

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/sql/scalar.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+            setattr(cls, meth, cls._get_comparison_operator(opt))
+
+    @classmethod
+    def _get_comparison_operator(
+        cls: Type["ComparisonOperatorsMinxin"], opt: str
+    ) -> Callable[..., Any]:
+        raise NotImplementedError
+
+
+class NumericalOperatorsMinxin(SearchContainer):
```
Comment (linjiX) [inline graviti/dataframe/sql/scalar.py:69]: Arithmetic

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/sql/frame.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    """The One-dimensional array for the search."""
+
+    def __init__(self, expr: str, schema: pt.record) -> None:
+        self.schema = schema
+        self.expr = expr
+
+    def __getitem__(self, key: str) -> Union["SearchContainer", "RowSeries", "DataFrame"]:
+        field = self.schema.to_builtin().fields[key]
+        expr = f"{self.expr}.{key}"
+        if isinstance(field, pt.record):
```
Comment (linjiX) [inline graviti/dataframe/sql/frame.py:27]: `field` may not be a builtin type

File: graviti/dataframe/sql/frame.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            return RowSeries(expr, field)
+
+        if isinstance(field, pt.array):
+            items = field.items
+            if isinstance(items, pt.record):
+                return DataFrame("$", items, expr)
+
+            if isinstance(items, (pt.int32, pt.int64, pt.float32, pt.float64)):
+                return ArrayNumericScalar("$.", field, expr)
```
Comment (linjiX) [inline graviti/dataframe/sql/frame.py:36]: why `$.` here but `$` in other place?

Comment (linjiX) [inline graviti/dataframe/sql/frame.py:36]: why `$.` here but `$` in other place?

File: graviti/dataframe/sql/scalar.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+                if NUMERICAL_PRIORITY[self.schema.__class__]
+                > NUMERICAL_PRIORITY[other.schema.__class__]
+                else other.schema
+            )
+            return cls(expr, schema)
+
+        return func
+
+    @staticmethod
+    def check_type_for_right(left: "NumberScalar", right: SearchContainer) -> None:
```
Comment (linjiX) [inline graviti/dataframe/sql/scalar.py:137]: shall we use `staticmethod` here? can it be a normal method?

Comment (linjiX) [inline graviti/dataframe/sql/scalar.py:137]: shall we use `staticmethod` here? can it be a normal method?

File: graviti/dataframe/sql/scalar.py [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+        Arguments:
+            left: The left series.
+            right: The right series.
+
+        Raises:
+            TypeError: When the right series type is different from left.
+
+        """
+        if not isinstance(right, left.__class__):
+            raise TypeError(f"Invalid operation between schema={left.schema} and {right.schema}")
```
Comment (linjiX) [inline graviti/dataframe/sql/scalar.py:284]: the error message should contain the operation type

Comment (linjiX) [inline graviti/dataframe/sql/scalar.py:284]: the error message should contain the operation type

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/sql/scalar.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+            if isinstance(other, SearchContainer):
+                expr = {f"${opt}": [self.expr, other.expr]}
+            else:
+                expr = {f"${opt}": [self.expr, other]}
+            return BooleanScalar(expr)
+
+        return func
+
+
+class ComparisonOperatorsMinxin:
```
Comment (linjiX) [inline graviti/dataframe/sql/scalar.py:53]: `Mixin`

File: graviti/dataframe/sql/scalar.py [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+    def __init_subclass__(cls) -> None:
+        super().__init_subclass__()
+        for meth, opt in cls.comparison_operators.items():
+            setattr(cls, meth, cls._get_comparison_operator(opt))
+
+    @classmethod
+    def _get_comparison_operator(cls, opt: str) -> Callable[..., Any]:
+        raise NotImplementedError
+
+    def check_type_for_other(self, other: SearchContainer, opt: str) -> None:
```
Comment (linjiX) [inline graviti/dataframe/sql/scalar.py:74]: this method is been defined twice, and the code is the same.

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/external.py [hunk 11]
```diff
@@ -51,6 +51,9 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         if not hasattr(self.__class__, "container"):
             self.container = self.internal_type.container
 
+        if not hasattr(self.__class__, "search_container"):
+            self.search_container = self.internal_type.search_container
```
Comment (linjiX) [inline graviti/portex/external.py:55]: set `internal_type` as a variable

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataframe/sql/array.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022 Graviti. Licensed under MIT License.
+#
+
+"""The implementation of the search DataFrame and Series."""
```
Comment (linjiX) [inline graviti/dataframe/sql/array.py:6]: The docstring should be updated

**Review by linjiX — CHANGES_REQUESTED** [review]

**Review by linjiX — APPROVED** [review]

File: graviti/dataframe/sql/array.py [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+    def _check_upper_expr(self, opt: str) -> None:
+        if self.expr != self.prefix:
+            raise TypeError(f"{opt} operation only support for numerical array.")
+
+    def size(self) -> "NumberScalar":
+        """Get the length of array series.
+
+        Returns:
+            The ArraySeries with the size expression.
```
Comment (Lee-000) [inline graviti/dataframe/sql/array.py:134]: `ArraySeries`?

File: graviti/dataframe/sql/array.py [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+    prefix = "$"
+    item_container = RowSeries
+
+    def __getitem__(self, key: str) -> SearchContainer:
+        field: pt.PortexType = self.schema.to_builtin().fields[key]  # type: ignore[attr-defined]
+        expr = f"{self.expr}.{key}"
+        return field.search_container(expr, field, self.upper_expr)
+
+    def query(self, func: Callable[[Any], Any]) -> "ScalarArray":
+        """Query the data of a ArraySeries with a lambda function.
```
Comment (Lee-000) [inline graviti/dataframe/sql/array.py:184]: `an ArraySeries`?

**Review by Lee-000 — APPROVED** [review]

**Pull request merged**